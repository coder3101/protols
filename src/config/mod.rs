use serde::{Deserialize, Serialize};

pub mod workspace;

fn default_clang_format_path() -> String {
    "clang-format".to_string()
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
#[serde(default)]
pub struct ProtolsConfig {
    pub config: Config,
    pub formatter: FormatterConfig,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct FormatterConfig {
    #[serde(default = "default_clang_format_path")]
    pub clang_format_path: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
#[serde(default)]
pub struct Config {
    pub include_paths: Vec<String>,
    pub single_file_mode: bool,
    pub disable_parse_diagnostics: bool,
    pub experimental: ExperimentalConfig,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
#[serde(default)]
pub struct ExperimentalConfig {
    pub use_protoc_diagnostics: bool,
}

impl Default for FormatterConfig {
    fn default() -> Self {
        Self {
            clang_format_path: default_clang_format_path(),
        }
    }
}
