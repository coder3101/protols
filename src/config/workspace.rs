use std::{
    collections::{HashMap, HashSet},
    env,
    path::{Path, PathBuf},
};

use async_lsp::lsp_types::{Url, WorkspaceFolder};
use pkg_config::Config;

use crate::formatter::clang::ClangFormatter;

use super::ProtolsConfig;

const CONFIG_FILE_NAMES: [&str; 2] = [".protols.toml", "protols.toml"];

pub struct WorkspaceProtoConfigs {
    workspaces: HashSet<Url>,
    configs: HashMap<Url, ProtolsConfig>,
    formatters: HashMap<Url, ClangFormatter>,
    protoc_include_prefix: Vec<PathBuf>,
    cli_include_paths: Vec<PathBuf>,
    init_include_paths: Vec<PathBuf>,
}

impl WorkspaceProtoConfigs {
    pub fn new(cli_include_paths: Vec<PathBuf>) -> Self {
        // Try to find protobuf library and get its include paths
        // Do not emit metadata on stdout as LSP programs can consider
        // it part of spec
        let protoc_include_prefix = Config::new()
            .atleast_version("3.0.0")
            .env_metadata(false)
            .cargo_metadata(false)
            .probe("protobuf")
            .map(|lib| lib.include_paths)
            .unwrap_or_default();

        Self {
            workspaces: HashSet::new(),
            formatters: HashMap::new(),
            configs: HashMap::new(),
            protoc_include_prefix,
            cli_include_paths,
            init_include_paths: Vec::new(),
        }
    }

    fn get_config_file_path(wpath: &PathBuf) -> Option<PathBuf> {
        for file in CONFIG_FILE_NAMES {
            let p = Path::new(&wpath).join(file);
            match std::fs::exists(&p) {
                Ok(exists) if exists => return Some(p),
                _ => continue,
            }
        }
        None
    }

    pub fn add_workspace(&mut self, w: &WorkspaceFolder) {
        let Ok(wpath) = w.uri.to_file_path() else {
            return;
        };

        let path = Self::get_config_file_path(&wpath).unwrap_or_default();
        let content = std::fs::read_to_string(path).unwrap_or_default();

        let wr: ProtolsConfig = basic_toml::from_str(&content).unwrap_or_default();
        let fmt = ClangFormatter::new(
            &wr.config.path.clang_format,
            Some(wpath.to_str().expect("non-utf8 path")),
        );

        self.workspaces.insert(w.uri.clone());
        self.configs.insert(w.uri.clone(), wr);
        self.formatters.insert(w.uri.clone(), fmt);
    }

    pub fn get_config_for_uri(&self, u: &Url) -> Option<&ProtolsConfig> {
        self.get_workspace_for_uri(u)
            .and_then(|w| self.configs.get(w))
    }

    pub fn get_formatter_for_uri(&self, u: &Url) -> Option<&ClangFormatter> {
        self.get_workspace_for_uri(u)
            .and_then(|w| self.formatters.get(w))
    }

    pub fn get_workspace_for_uri(&self, u: &Url) -> Option<&Url> {
        let upath = u.to_file_path().ok()?;
        self.workspaces
            .iter()
            .find(|&k| upath.starts_with(k.to_file_path().unwrap()))
    }

    pub fn set_init_include_paths(&mut self, paths: Vec<PathBuf>) {
        self.init_include_paths = paths;
    }

    pub fn get_include_paths(&self, uri: &Url) -> Option<Vec<PathBuf>> {
        let cfg = self.get_config_for_uri(uri)?;
        let w = self.get_workspace_for_uri(uri)?.to_file_path().ok()?;

        let mut ipath: Vec<PathBuf> = cfg
            .config
            .include_paths
            .iter()
            .map(PathBuf::from)
            .map(|p| if p.is_relative() { w.join(p) } else { p })
            .collect();

        // Add CLI include paths
        for path in &self.cli_include_paths {
            if path.is_relative() {
                ipath.push(w.join(path));
            } else {
                ipath.push(path.clone());
            }
        }

        // Add initialization include paths
        for path in &self.init_include_paths {
            if path.is_relative() {
                ipath.push(w.join(path));
            } else {
                ipath.push(path.clone());
            }
        }

        ipath.push(w.to_path_buf());
        ipath.extend_from_slice(&self.protoc_include_prefix);
        Some(ipath)
    }

    pub fn no_workspace_mode(&mut self) {
        let wr = ProtolsConfig::default();
        let rp = if cfg!(target_os = "windows") {
            let mut d = String::from("C");
            if let Ok(cdir) = env::current_dir()
                && let Some(drive) = cdir.components().next()
            {
                d = drive.as_os_str().to_string_lossy().to_string()
            }
            format!("{d}://")
        } else {
            String::from("/")
        };
        let uri = match Url::from_file_path(&rp) {
            Err(err) => {
                tracing::error!(?err, "failed to convert path: {rp} to Url");
                return;
            }
            Ok(uri) => uri,
        };

        let fmt = ClangFormatter::new(&wr.config.path.clang_format, None);

        self.workspaces.insert(uri.clone());
        self.configs.insert(uri.clone(), wr);
        self.formatters.insert(uri.clone(), fmt);
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::{Url, WorkspaceFolder};
    use insta::assert_yaml_snapshot;
    use std::path::PathBuf;
    use tempfile::tempdir;

    use super::{CONFIG_FILE_NAMES, WorkspaceProtoConfigs};

    #[test]
    fn test_get_for_workspace() {
        let tmpdir = tempdir().expect("failed to create temp directory");
        let tmpdir2 = tempdir().expect("failed to create temp2 directory");
        let f = tmpdir.path().join("protols.toml");
        std::fs::write(f, include_str!("input/protols-valid.toml")).unwrap();

        let mut ws = WorkspaceProtoConfigs::new(vec![]);
        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir.path()).unwrap(),
            name: "Test".to_string(),
        });
        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir2.path()).unwrap(),
            name: "Test2".to_string(),
        });

        let inworkspace = Url::from_file_path(tmpdir.path().join("foobar.proto")).unwrap();
        let outworkspace =
            Url::from_file_path(tempdir().unwrap().path().join("out.proto")).unwrap();
        let inworkspace2 = Url::from_file_path(tmpdir2.path().join("foobar.proto")).unwrap();

        assert!(ws.get_config_for_uri(&inworkspace).is_some());
        assert!(ws.get_config_for_uri(&inworkspace2).is_some());
        assert!(ws.get_config_for_uri(&outworkspace).is_none());

        assert!(ws.get_workspace_for_uri(&inworkspace).is_some());
        assert!(ws.get_workspace_for_uri(&inworkspace2).is_some());
        assert!(ws.get_workspace_for_uri(&outworkspace).is_none());

        assert_yaml_snapshot!(ws.get_config_for_uri(&inworkspace).unwrap());
        assert_yaml_snapshot!(ws.get_config_for_uri(&inworkspace2).unwrap());
    }

    #[test]
    fn test_get_formatter_for_uri() {
        let tmpdir = tempdir().expect("failed to create temp directory");
        let tmpdir2 = tempdir().expect("failed to create temp2 directory");
        let f = tmpdir.path().join("protols.toml");
        std::fs::write(f, include_str!("input/protols-valid.toml")).unwrap();

        let mut ws = WorkspaceProtoConfigs::new(vec![]);
        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir.path()).unwrap(),
            name: "Test".to_string(),
        });

        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir2.path()).unwrap(),
            name: "Test2".to_string(),
        });

        let inworkspace = Url::from_file_path(tmpdir.path().join("foobar.proto")).unwrap();
        let outworkspace =
            Url::from_file_path(tempdir().unwrap().path().join("out.proto")).unwrap();
        let inworkspace2 = Url::from_file_path(tmpdir2.path().join("foobar.proto")).unwrap();

        assert!(ws.get_formatter_for_uri(&outworkspace).is_none());
        assert_eq!(
            ws.get_formatter_for_uri(&inworkspace).unwrap().path,
            "/usr/bin/clang-format"
        );
        assert_eq!(
            ws.get_formatter_for_uri(&inworkspace2).unwrap().path,
            "clang-format"
        );
    }

    #[test]
    fn test_loading_different_config_files() {
        let tmpdir = tempdir().expect("failed to create temp directory");

        for file in CONFIG_FILE_NAMES {
            let f = tmpdir.path().join(file);
            std::fs::write(f, include_str!("input/protols-valid.toml")).unwrap();

            let mut ws = WorkspaceProtoConfigs::new(vec![]);
            ws.add_workspace(&WorkspaceFolder {
                uri: Url::from_directory_path(tmpdir.path()).unwrap(),
                name: "Test".to_string(),
            });

            // check we really loaded the config file
            let workspace = Url::from_file_path(tmpdir.path().join("foobar.proto")).unwrap();
            assert!(ws.get_workspace_for_uri(&workspace).is_some());
        }
    }

    #[test]
    fn test_cli_include_paths() {
        let tmpdir = tempdir().expect("failed to create temp directory");
        let f = tmpdir.path().join("protols.toml");
        std::fs::write(f, include_str!("input/protols-valid.toml")).unwrap();

        // Set CLI include paths
        let cli_paths = vec![
            PathBuf::from("/path/to/protos"),
            PathBuf::from("relative/path"),
        ];
        let mut ws = WorkspaceProtoConfigs::new(cli_paths);
        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir.path()).unwrap(),
            name: "Test".to_string(),
        });

        let inworkspace = Url::from_file_path(tmpdir.path().join("foobar.proto")).unwrap();
        let include_paths = ws.get_include_paths(&inworkspace).unwrap();

        // Check that CLI paths are included in the result
        assert!(
            include_paths
                .iter()
                .any(|p| p.ends_with("relative/path") || p == &PathBuf::from("/path/to/protos"))
        );

        // The relative path should be resolved relative to the workspace
        let resolved_relative_path = tmpdir.path().join("relative/path");
        assert!(include_paths.contains(&resolved_relative_path));

        // The absolute path should be included as is
        assert!(include_paths.contains(&PathBuf::from("/path/to/protos")));
    }

    #[test]
    fn test_init_include_paths() {
        let tmpdir = tempdir().expect("failed to create temp directory");
        let f = tmpdir.path().join("protols.toml");
        std::fs::write(f, include_str!("input/protols-valid.toml")).unwrap();

        // Set both CLI and initialization include paths
        let cli_paths = vec![PathBuf::from("/cli/path")];
        let init_paths = vec![
            PathBuf::from("/init/path1"),
            PathBuf::from("relative/init/path"),
        ];

        let mut ws = WorkspaceProtoConfigs::new(cli_paths);
        ws.set_init_include_paths(init_paths);
        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir.path()).unwrap(),
            name: "Test".to_string(),
        });

        let inworkspace = Url::from_file_path(tmpdir.path().join("foobar.proto")).unwrap();
        let include_paths = ws.get_include_paths(&inworkspace).unwrap();

        // Check that initialization paths are included
        assert!(include_paths.contains(&PathBuf::from("/init/path1")));

        // The relative path should be resolved relative to the workspace
        let resolved_relative_path = tmpdir.path().join("relative/init/path");
        assert!(include_paths.contains(&resolved_relative_path));

        // CLI paths should still be included
        assert!(include_paths.contains(&PathBuf::from("/cli/path")));
    }
}
