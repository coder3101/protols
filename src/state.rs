use std::{
    collections::HashMap,
    fs::read_to_string,
    sync::{mpsc::Sender, Arc, Mutex, MutexGuard, RwLock, RwLockWriteGuard},
    thread,
};
use tracing::{error, info};

use async_lsp::lsp_types::{
    CompletionItem, CompletionItemKind, ProgressParamsValue, PublishDiagnosticsParams, Url,
    WorkDoneProgress, WorkDoneProgressBegin, WorkDoneProgressEnd, WorkDoneProgressReport,
    WorkspaceFolder,
};
use tree_sitter::Node;
use walkdir::WalkDir;

use crate::{
    nodekind::NodeKind,
    parser::{ParsedTree, ProtoParser},
};

pub struct ProtoLanguageState {
    documents: Arc<RwLock<HashMap<Url, String>>>,
    trees: Arc<RwLock<HashMap<Url, ParsedTree>>>,
    parser: Arc<Mutex<ProtoParser>>,
}

impl ProtoLanguageState {
    pub fn new() -> Self {
        ProtoLanguageState {
            documents: Default::default(),
            trees: Default::default(),
            parser: Arc::new(Mutex::new(ProtoParser::new())),
        }
    }

    pub fn get_content(&self, uri: &Url) -> String {
        self.documents
            .read()
            .expect("poison")
            .get(uri)
            .map(|s| s.to_string())
            .unwrap_or_default()
    }

    pub fn get_tree(&self, uri: &Url) -> Option<ParsedTree> {
        self.trees.read().expect("poison").get(uri).cloned()
    }

    pub fn get_trees(&self) -> Vec<ParsedTree> {
        self.trees
            .read()
            .expect("poison")
            .values()
            .map(ToOwned::to_owned)
            .collect()
    }

    pub fn get_trees_for_package(&self, package: &str) -> Vec<ParsedTree> {
        self.trees
            .read()
            .expect("poison")
            .values()
            .filter(|tree| {
                let content = self.get_content(&tree.uri);
                tree.get_package_name(content.as_bytes())
                    .unwrap_or_default()
                    == package
            })
            .map(ToOwned::to_owned)
            .collect()
    }

    fn upsert_content_impl(
        mut parser: MutexGuard<ProtoParser>,
        uri: &Url,
        content: String,
        mut docs: RwLockWriteGuard<HashMap<Url, String>>,
        mut trees: RwLockWriteGuard<HashMap<Url, ParsedTree>>,
    ) -> bool {
        if let Some(parsed) = parser.parse(uri.clone(), content.as_bytes()) {
            trees.insert(uri.clone(), parsed);
            docs.insert(uri.clone(), content);
            true
        } else {
            false
        }
    }

    pub fn upsert_content(&mut self, uri: &Url, content: String) -> bool {
        let parser = self.parser.lock().expect("poison");
        let tree = self.trees.write().expect("poison");
        let docs = self.documents.write().expect("poison");
        Self::upsert_content_impl(parser, uri, content, docs, tree)
    }

    pub fn add_workspace_folder_async(
        &mut self,
        workspace: WorkspaceFolder,
        tx: Sender<ProgressParamsValue>,
    ) {
        let parser = self.parser.clone();
        let tree = self.trees.clone();
        let docs = self.documents.clone();

        let begin = ProgressParamsValue::WorkDone(WorkDoneProgress::Begin(WorkDoneProgressBegin {
            title: String::from("indexing"),
            cancellable: Some(false),
            percentage: Some(0),
            ..Default::default()
        }));

        if let Err(e) = tx.send(begin) {
            error!(error=%e, "failed to send work begin progress");
        }

        thread::spawn(move || {
            let files: Vec<_> = WalkDir::new(workspace.uri.path())
                .into_iter()
                .filter_map(|e| e.ok())
                .filter(|e| e.path().extension().is_some())
                .filter(|e| e.path().extension().unwrap() == "proto")
                .collect();

            let total_files = files.len();
            let mut current = 0;

            for file in files.into_iter() {
                let path = file.path();
                if path.is_absolute() && path.is_file() {
                    let Ok(content) = read_to_string(path) else {
                        continue;
                    };

                    let Ok(uri) = Url::from_file_path(path) else {
                        continue;
                    };

                    Self::upsert_content_impl(
                        parser.lock().expect("poison"),
                        &uri,
                        content,
                        docs.write().expect("poison"),
                        tree.write().expect("poison"),
                    );

                    current += 1;

                    let report = ProgressParamsValue::WorkDone(WorkDoneProgress::Report(
                        WorkDoneProgressReport {
                            cancellable: Some(false),
                            message: Some(path.display().to_string()),
                            percentage: Some((current * 100 / total_files) as u32),
                        },
                    ));

                    if let Err(e) = tx.send(report) {
                        error!(error=%e, "failed to send work report progress");
                    }
                }
            }
            let report =
                ProgressParamsValue::WorkDone(WorkDoneProgress::End(WorkDoneProgressEnd {
                    message: Some(String::from("completed")),
                }));

            info!(len = total_files, "workspace file parsing completed");
            if let Err(e) = tx.send(report) {
                error!(error=%e, "failed to send work completed result");
            }
        });
    }

    pub fn upsert_file(&mut self, uri: &Url, content: String) -> Option<PublishDiagnosticsParams> {
        info!(uri=%uri, "upserting file");
        self.upsert_content(uri, content);
        self.get_tree(uri).map(|tree| tree.collect_parse_errors())
    }

    pub fn delete_file(&mut self, uri: &Url) {
        info!(uri=%uri, "deleting file");
        self.documents.write().expect("poison").remove(uri);
        self.trees.write().expect("poison").remove(uri);
    }

    pub fn rename_file(&mut self, new_uri: &Url, old_uri: &Url) {
        info!(new_uri=%new_uri, old_uri=%new_uri, "renaming file");

        if let Some(v) = self.documents.write().expect("poison").remove(old_uri) {
            self.documents
                .write()
                .expect("poison")
                .insert(new_uri.clone(), v);
        }

        if let Some(mut v) = self.trees.write().expect("poison").remove(old_uri) {
            v.uri = new_uri.clone();
            self.trees
                .write()
                .expect("poison")
                .insert(new_uri.clone(), v);
        }
    }

    pub fn completion_items(&self, package: &str) -> Vec<CompletionItem> {
        let collector = |f: fn(&Node) -> bool, k: CompletionItemKind| {
            self.get_trees_for_package(package)
                .into_iter()
                .fold(vec![], |mut v, tree| {
                    let content = self.get_content(&tree.uri);
                    let t = tree.filter_nodes(f).into_iter().map(|n| CompletionItem {
                        label: n.utf8_text(content.as_bytes()).unwrap().to_string(),
                        kind: Some(k),
                        ..Default::default()
                    });
                    v.extend(t);
                    v
                })
        };

        let mut result = collector(NodeKind::is_enum_name, CompletionItemKind::ENUM);
        result.extend(collector(
            NodeKind::is_message_name,
            CompletionItemKind::STRUCT,
        ));
        // Better ways to dedup, but who cares?...
        result.sort_by_key(|k| k.label.clone());
        result.dedup_by_key(|k| k.label.clone());
        result
    }
}
