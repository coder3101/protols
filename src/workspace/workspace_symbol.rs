#[cfg(test)]
mod test {
    use insta::assert_yaml_snapshot;

    use crate::config::Config;
    use crate::state::ProtoLanguageState;

    #[test]
    fn test_workspace_symbols() {
        let ipath = vec![std::env::current_dir().unwrap().join("src/workspace/input")];
        let a_uri = "file://input/a.proto".parse().unwrap();
        let b_uri = "file://input/b.proto".parse().unwrap();
        let c_uri = "file://input/c.proto".parse().unwrap();

        let a = include_str!("input/a.proto");
        let b = include_str!("input/b.proto");
        let c = include_str!("input/c.proto");

        let mut state: ProtoLanguageState = ProtoLanguageState::new();
        state.upsert_file(&a_uri, a.to_owned(), &ipath, 3, &Config::default(), false);
        state.upsert_file(&b_uri, b.to_owned(), &ipath, 2, &Config::default(), false);
        state.upsert_file(&c_uri, c.to_owned(), &ipath, 2, &Config::default(), false);

        // Test empty query - should return all symbols
        let all_symbols = state.find_workspace_symbols("");
        assert_yaml_snapshot!("all_symbols", all_symbols);

        // Test query for "author" - should match Author and Address
        let author_symbols = state.find_workspace_symbols("author");
        assert_yaml_snapshot!("author_symbols", author_symbols);

        // Test query for "address" - should match Address
        let address_symbols = state.find_workspace_symbols("address");
        assert_yaml_snapshot!("address_symbols", address_symbols);

        // Test query that should not match anything
        let no_match = state.find_workspace_symbols("nonexistent");
        assert!(no_match.is_empty());
    }
}
