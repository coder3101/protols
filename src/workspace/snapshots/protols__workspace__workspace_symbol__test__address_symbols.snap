---
source: src/workspace/workspace_symbol.rs
expression: address_symbols
---
- name: Address
  kind: 23
  containerName: Author
  location:
    uri: "file:///home/runner/work/protols/protols/src/workspace/input/b.proto"
    range:
      start:
        line: 9
        character: 3
      end:
        line: 11
        character: 4
- name: Address
  kind: 23
  containerName: Author
  location:
    uri: "file://input/b.proto"
    range:
      start:
        line: 9
        character: 3
      end:
        line: 11
        character: 4
